lib/web_ui/lib/src/engine/text/ruler.dart

```diff
@@ -628,6 +628,27 @@ class ParagraphRuler {
     );
   }
 
+  List<ui.TextBox> measurePlaceholderBoxes() {
+    assert(!_debugIsDisposed);
+    assert(_paragraph != null);
+
+    final List<html.Element> placeholderElements =
+        constrainedDimensions._element.querySelectorAll('.$_placeholderClass');
```

Comment:
Is this fast enough? How does it compare to storing the placeholders in our own list?

Reply:
Our own list won't work because the ruler will deeply copy the paragraph element, including the placeholder children. So we need to re-query for placeholder elements within the ruler's clone.

One optimization I can do here is check if the number of placeholders in the paragraph is 0 then shortcircuit this method and skip the query selector.

Reply:
Ah, gotcha.